In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# load the the worldbank dataset 
world_bank_data = pd.read_csv("data/World_Bank_Data.csv")

In [3]:
world_bank_data.head(20)

Country Name Country Code  \
0   Afghanistan          AFG   
1   Afghanistan          AFG   
2   Afghanistan          AFG   
3   Afghanistan          AFG   
4   Afghanistan          AFG   
5   Afghanistan          AFG   
6   Afghanistan          AFG   
7   Afghanistan          AFG   
8   Afghanistan          AFG   
9   Afghanistan          AFG   
10  Afghanistan          AFG   
11      Albania          ALB   
12      Albania          ALB   
13      Albania          ALB   
14      Albania          ALB   
15      Albania          ALB   
16      Albania          ALB   
17      Albania          ALB   
18      Albania          ALB   
19      Albania          ALB   

                                          Series Name        Series Code  \
0                        GDP per capita (current US$)     NY.GDP.PCAP.CD   
1               Current health expenditure (% of GDP)  SH.XPD.CHEX.GD.ZS   
2   Current health expenditure per capita (current...  SH.XPD.CHEX.PC.CD   
3   Current education expenditure, total (% of tot...     SE.XPD.CTOT.ZS   
4             Life expectancy at birth, total (years)     SP.DYN.LE00.IN   
5                                   Population, total        SP.POP.TOTL   
6            Fertility rate, total (births per woman)     SP.DYN.TFRT.IN   
7      Mortality rate, infant (per 1,000 live births)     SP.DYN.IMRT.IN   
8   Age dependency ratio (% of working-age populat...        SP.POP.DPND   
9   PM2.5 air pollution, mean annual exposure (mic...  EN.ATM.PM25.MC.M3   
10  Carbon dioxide (CO2) emissions (total) excludi...  EN.GHG.CO2.ZG.AR5   
11                       GDP per capita (current US$)     NY.GDP.PCAP.CD   
12              Current health expenditure (% of GDP)  SH.XPD.CHEX.GD.ZS   
13  Current health expenditure per capita (current...  SH.XPD.CHEX.PC.CD   
14  Current education expenditure, total (% of tot...     SE.XPD.CTOT.ZS   
15            Life expectancy at birth, total (years)     SP.DYN.LE00.IN   
16                                  Population, total        SP.POP.TOTL   
17           Fertility rate, total (births per woman)     SP.DYN.TFRT.IN   
18     Mortality rate, infant (per 1,000 live births)     SP.DYN.IMRT.IN   
19  Age dependency ratio (% of working-age populat...        SP.POP.DPND   

       1960 [YR1960]     1961 [YR1961]     1962 [YR1962]     1963 [YR1963]  \
0                 ..                ..                ..                ..   
1                 ..                ..                ..                ..   
2                 ..                ..                ..                ..   
3                 ..                ..                ..                ..   
4             32.799            33.291            33.757            34.201   
5            9035043           9214083           9404406           9604487   
6              7.282             7.284             7.292             7.302   
7              251.2             248.4             245.4             242.5   
8   81.0613894386502  81.2840359519206  81.5068083352119  81.8545805002831   
9                 ..                ..                ..                ..   
10                ..                ..                ..                ..   
11                ..                ..                ..                ..   
12                ..                ..                ..                ..   
13                ..                ..                ..                ..   
14                ..                ..                ..                ..   
15            56.413            57.488            58.494            59.479   
16           1608800           1659800           1711319           1762621   
17             6.383             6.273             6.106             5.927   
18                ..                ..                ..                ..   
19  87.4824468277266  89.0174463626839  89.6692213484952  89.8996428513712   

       1964 [YR1964]     1965 [YR1965]  ...     2015 [YR2015]  \
0                 ..         

The data is in wide format. All the years and indicators data are spread across the data frame. We need to restructure the indicator data and year into long format. 

Which mean we will change the following data wide format 

Country Name | Country Code | Series Name | Series Code | 1960 | 1961 | .... | Value 

into 

Country Name | Country Code | Series Name | Series Code | Year | Value 

In [4]:
world_bank_data_long = world_bank_data.melt(id_vars=["Country Name","Country Code", "Series Name", "Series Code"], 
                  var_name="Year", value_name="Value")


In [5]:
world_bank_data_long.head(20)

Country Name Country Code  \
0   Afghanistan          AFG   
1   Afghanistan          AFG   
2   Afghanistan          AFG   
3   Afghanistan          AFG   
4   Afghanistan          AFG   
5   Afghanistan          AFG   
6   Afghanistan          AFG   
7   Afghanistan          AFG   
8   Afghanistan          AFG   
9   Afghanistan          AFG   
10  Afghanistan          AFG   
11      Albania          ALB   
12      Albania          ALB   
13      Albania          ALB   
14      Albania          ALB   
15      Albania          ALB   
16      Albania          ALB   
17      Albania          ALB   
18      Albania          ALB   
19      Albania          ALB   

                                          Series Name        Series Code  \
0                        GDP per capita (current US$)     NY.GDP.PCAP.CD   
1               Current health expenditure (% of GDP)  SH.XPD.CHEX.GD.ZS   
2   Current health expenditure per capita (current...  SH.XPD.CHEX.PC.CD   
3   Current education expenditure, total (% of tot...     SE.XPD.CTOT.ZS   
4             Life expectancy at birth, total (years)     SP.DYN.LE00.IN   
5                                   Population, total        SP.POP.TOTL   
6            Fertility rate, total (births per woman)     SP.DYN.TFRT.IN   
7      Mortality rate, infant (per 1,000 live births)     SP.DYN.IMRT.IN   
8   Age dependency ratio (% of working-age populat...        SP.POP.DPND   
9   PM2.5 air pollution, mean annual exposure (mic...  EN.ATM.PM25.MC.M3   
10  Carbon dioxide (CO2) emissions (total) excludi...  EN.GHG.CO2.ZG.AR5   
11                       GDP per capita (current US$)     NY.GDP.PCAP.CD   
12              Current health expenditure (% of GDP)  SH.XPD.CHEX.GD.ZS   
13  Current health expenditure per capita (current...  SH.XPD.CHEX.PC.CD   
14  Current education expenditure, total (% of tot...     SE.XPD.CTOT.ZS   
15            Life expectancy at birth, total (years)     SP.DYN.LE00.IN   
16                                  Population, total        SP.POP.TOTL   
17           Fertility rate, total (births per woman)     SP.DYN.TFRT.IN   
18     Mortality rate, infant (per 1,000 live births)     SP.DYN.IMRT.IN   
19  Age dependency ratio (% of working-age populat...        SP.POP.DPND   

             Year             Value  
0   1960 [YR1960]                ..  
1   1960 [YR1960]                ..  
2   1960 [YR1960]                ..  
3   1960 [YR1960]                ..  
4   1960 [YR1960]            32.799  
5   1960 [YR1960]           9035043  
6   1960 [YR1960]             7.282  
7   1960 [YR1960]             251.2  
8   1960 [YR1960]  81.0613894386502  
9   1960 [YR1960]                ..  
10  1960 [YR1960]                ..  
11  1960 [YR1960]                ..  
12  1960 [YR1960]                ..  
13  1960 [YR1960]                ..  
14  1960 [YR1960]                ..  
15  1960 [YR1960]            56.413  
16  1960 [YR1960]           1608800  
17  1960 [YR1960]             6.383  
18  1960 [YR1960]                ..  
19  1960 [YR1960]  87.4824468277266

In [6]:
# world_bank_data_long[world_bank_data_long['Country Name'] == 'Afghanistan']
df_world = world_bank_data_long.copy()

In [7]:
# check the info of the dataframe
df_world.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 190515 entries, 0 to 190514
Data columns (total 6 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   Country Name  190320 non-null  object
 1   Country Code  190190 non-null  object
 2   Series Name   190190 non-null  object
 3   Series Code   190190 non-null  object
 4   Year          190515 non-null  object
 5   Value         190190 non-null  object
dtypes: object(6)
memory usage: 8.7+ MB


In [8]:
# check the basic statics
df_world.describe()

Country Name Country Code                   Series Name  \
count        190320       190190                        190190   
unique          268          266                            11   
top     Afghanistan          AFG  GDP per capita (current US$)   
freq            715          715                         17290   

           Series Code           Year   Value  
count           190190         190515  190190  
unique              11             65  102972  
top     NY.GDP.PCAP.CD  1960 [YR1960]      ..  
freq             17290           2931   65409

In [9]:
# print out the first 3 rows 
df_world.head(3)

Country Name Country Code  \
0  Afghanistan          AFG   
1  Afghanistan          AFG   
2  Afghanistan          AFG   

                                         Series Name        Series Code  \
0                       GDP per capita (current US$)     NY.GDP.PCAP.CD   
1              Current health expenditure (% of GDP)  SH.XPD.CHEX.GD.ZS   
2  Current health expenditure per capita (current...  SH.XPD.CHEX.PC.CD   

            Year Value  
0  1960 [YR1960]    ..  
1  1960 [YR1960]    ..  
2  1960 [YR1960]    ..

In the Year column, the year are in format like 1960 [YR1960]. We only want 1960 as year and in int data type. We will extract those values from the Year column.

In [10]:
# df_world['Year'] = df_world['Year'].str.extract(r'\d{4}').astype(int)
# extrac the year from the Year column and change to integer data type
df_world['Year'] = df_world['Year'].str.extract(r'(\d{4})').astype(int)
df_world.head(3)

Country Name Country Code  \
0  Afghanistan          AFG   
1  Afghanistan          AFG   
2  Afghanistan          AFG   

                                         Series Name        Series Code  Year  \
0                       GDP per capita (current US$)     NY.GDP.PCAP.CD  1960   
1              Current health expenditure (% of GDP)  SH.XPD.CHEX.GD.ZS  1960   
2  Current health expenditure per capita (current...  SH.XPD.CHEX.PC.CD  1960   

  Value  
0    ..  
1    ..  
2    ..

In [11]:
# check the Year column data type
df_world['Year'].dtype

dtype('int64')

In addition, inside Value column the missing values are stated as ".." which is not a good format. These are error prone place holders and will lead to issue when we tried to train the model. Therefore, we will replace them with NaN then we will change to numeric because most of the values are numeric. 

In [12]:
# replace the .. with NA
df_world['Value'] = df_world['Value'].replace('..', pd.NA)
# change the value column to numerical values
df_world['Value'] = pd.to_numeric(df_world['Value'], errors='coerce')

In [13]:
# check the first 3 rows of Value column
df_world['Value'].head(3)

0   NaN
1   NaN
2   NaN
Name: Value, dtype: float64

In [14]:
# check the value column datatype
df_world.Value.dtype

dtype('float64')

In [15]:
# display the first 5 rows
df_world.head(5)

Country Name Country Code  \
0  Afghanistan          AFG   
1  Afghanistan          AFG   
2  Afghanistan          AFG   
3  Afghanistan          AFG   
4  Afghanistan          AFG   

                                         Series Name        Series Code  Year  \
0                       GDP per capita (current US$)     NY.GDP.PCAP.CD  1960   
1              Current health expenditure (% of GDP)  SH.XPD.CHEX.GD.ZS  1960   
2  Current health expenditure per capita (current...  SH.XPD.CHEX.PC.CD  1960   
3  Current education expenditure, total (% of tot...     SE.XPD.CTOT.ZS  1960   
4            Life expectancy at birth, total (years)     SP.DYN.LE00.IN  1960   

    Value  
0     NaN  
1     NaN  
2     NaN  
3     NaN  
4  32.799

In [16]:
# check the rows and columns
df_world.shape

(190515, 6)

Currently, the values and years are in each line which is not so favourable. For example, there are 11 indicators collected now and in each year there will be 11 values for each indicators. If the country's data is accessible from 1960 to 2025 there will be 715 rows for a country. There are 266 countries and the no. of rows will be explode to 190,190 which is not so feasiable. Therefore, we need to transform the data into columnar format. 

From this - 

| Country Name | Country Code | Series Name | Series Code | Year | Value |
| ------------ | ------------ | ----------- | ----------- | ---- | ----- |
| Afghanistan  | AFG          | GDP per capita (current US$)| NY.GDP.PCAP.CD | 1960 | NaN |
| Afghanistan  | AFG          | Current health expenditure (% of GDP) | SH.XPD.CHEX.GD.ZS | 1960 | NaN |

To this -

| Country Name | Country Code | Year | GDP per capita | Current health expenditure |
| ------------ | ------------ | ---- | -------------- | -------------------------- |
| Afghanistan  | AFG          | 1960 | NaN            | NaN                        |

Note: the remaining rows will be transformed into columns as above.

In [17]:
# select indicators to kee 
keep_codes = [
    'SP.DYN.LE00.IN',          # Life expectancy at birth, total (years)
    'NY.GDP.PCAP.CD',          # GDP per capita (current US$)
    'SH.XPD.CHEX.GD.ZS',       # Current health expenditure (% of GDP)
    'SH.XPD.CHEX.PC.CD',       # Current health expenditure per capita (current US$)
    'SE.XPD.CTOT.ZS',          # Current education expenditure, total (% of total expenditure in public institutions)
    'SP.POP.TOTL',             # Population, total
    'SP.DYN.TFRT.IN',          # Fertility rate, total (births per woman)
    'SP.DYN.IMRT.IN',          # Mortality rate, infant (per 1,000 live births)
    'SP.POP.DPND',             # Age dependency ratio (% of working-age population)
    'EN.ATM.PM25.MC.M3',       # PM2.5 air pollution, mean annual exposure (micrograms per cubic meter)
    'EN.ATM.CO2E.PC',          # Carbon dioxide (CO2) emissions (total) excluding LULUCF (% change from 1990)
]
# extract the inicators
df_world = df_world[df_world['Series Code'].isin(keep_codes)]


In [18]:
# pivot the series into columnar format
df_world_wide = (
    df_world.pivot_table(
        index=['Country Name','Country Code','Year'],
        columns='Series Code',
        values='Value',
    )
    .reset_index()
)

In [19]:
# name mapping
name_map = {
    'SP.DYN.LE00.IN': 'life_expectancy',
    'NY.GDP.PCAP.CD': 'gdp_per_capita_usd',
    'SH.XPD.CHEX.GD.ZS': 'health_exp_pct_gdp',
    'SH.XPD.CHEX.PC.CD': 'health_exp_pc_usd',
    'SE.XPD.CTOT.ZS': 'edu_exp_pct_public',
    'SP.POP.TOTL': 'population',
    'SP.DYN.TFRT.IN': 'fertility_rate',
    'SP.DYN.IMRT.IN': 'infant_mortality',
    'SP.POP.DPND': 'age_dependency_ratio',
    'EN.ATM.PM25.MC.M3': 'pm_2_5',
    'EN.ATM.CO2E.PC': 'co2_pc',
}

In [20]:
# sort the index with country name and year
df_world = df_world_wide.rename(columns=name_map).sort_values(['Country Name','Year'])

In [21]:
# display the first 5 rows
df_world.head(5)

Series Code Country Name Country Code  Year  pm_2_5  gdp_per_capita_usd  \
0            Afghanistan          AFG  1960     NaN                 NaN   
1            Afghanistan          AFG  1961     NaN                 NaN   
2            Afghanistan          AFG  1962     NaN                 NaN   
3            Afghanistan          AFG  1963     NaN                 NaN   
4            Afghanistan          AFG  1964     NaN                 NaN   

Series Code  edu_exp_pct_public  health_exp_pct_gdp  health_exp_pc_usd  \
0                           NaN                 NaN                NaN   
1                           NaN                 NaN                NaN   
2                           NaN                 NaN                NaN   
3                           NaN                 NaN                NaN   
4                           NaN                 NaN                NaN   

Series Code  infant_mortality  life_expectancy  fertility_rate  \
0                       251.2           32.799           7.282   
1                       248.4           33.291           7.284   
2                       245.4           33.757           7.292   
3                       242.5           34.201           7.302   
4                       239.7           34.673           7.304   

Series Code  age_dependency_ratio  population  
0                       81.061389   9035043.0  
1                       81.284036   9214083.0  
2                       81.506808   9404406.0  
3                       81.854581   9604487.0  
4                       82.373013   9814318.0

In [22]:
df_world.tail()

Series Code Country Name Country Code  Year    pm_2_5  gdp_per_capita_usd  \
17205           Zimbabwe          ZWE  2020  19.49418         1730.453910   
17206           Zimbabwe          ZWE  2021       NaN         1724.387271   
17207           Zimbabwe          ZWE  2022       NaN         2040.546587   
17208           Zimbabwe          ZWE  2023       NaN         2156.034093   
17209           Zimbabwe          ZWE  2024       NaN         2656.409377   

Series Code  edu_exp_pct_public  health_exp_pct_gdp  health_exp_pc_usd  \
17205                       NaN            2.954401          51.142506   
17206                       NaN            2.785717          63.511448   
17207                       NaN            3.628807          70.663139   
17208                       NaN                 NaN                NaN   
17209                       NaN                 NaN                NaN   

Series Code  infant_mortality  life_expectancy  fertility_rate  \
17205                    44.9           61.530           3.754   
17206                    43.6           60.135           3.765   
17207                    42.2           62.360           3.767   
17208                    40.6           62.775           3.724   
17209                     NaN              NaN             NaN   

Series Code  age_dependency_ratio  population  
17205                   84.384381  15526888.0  
17206                   83.384953  15797210.0  
17207                   82.547605  16069056.0  
17208                   81.625352  16340822.0  
17209                   80.100412  16634373.0

In [23]:
# rename the index
df_world.columns.name = None

In [24]:
# rename the first three columns into lower case
df_world = df_world.rename(columns={
    'Country Name': 'country_name',
    'Country Code': 'country_code',
    'Year': 'year'
})


In [25]:
# check the rows and columns
df_world.shape

(17210, 13)

In [26]:
# display the first 5 rows
df_world.head(5)

country_name country_code  year  pm_2_5  gdp_per_capita_usd  \
0  Afghanistan          AFG  1960     NaN                 NaN   
1  Afghanistan          AFG  1961     NaN                 NaN   
2  Afghanistan          AFG  1962     NaN                 NaN   
3  Afghanistan          AFG  1963     NaN                 NaN   
4  Afghanistan          AFG  1964     NaN                 NaN   

   edu_exp_pct_public  health_exp_pct_gdp  health_exp_pc_usd  \
0                 NaN                 NaN                NaN   
1                 NaN                 NaN                NaN   
2                 NaN                 NaN                NaN   
3                 NaN                 NaN                NaN   
4                 NaN                 NaN                NaN   

   infant_mortality  life_expectancy  fertility_rate  age_dependency_ratio  \
0             251.2           32.799           7.282             81.061389   
1             248.4           33.291           7.284             81.284036   
2             245.4           33.757           7.292             81.506808   
3             242.5           34.201           7.302             81.854581   
4             239.7           34.673           7.304             82.373013   

   population  
0   9035043.0  
1   9214083.0  
2   9404406.0  
3   9604487.0  
4   9814318.0

In [27]:
# display the last 5 rows 
df_world.tail(5)

country_name country_code  year    pm_2_5  gdp_per_capita_usd  \
17205     Zimbabwe          ZWE  2020  19.49418         1730.453910   
17206     Zimbabwe          ZWE  2021       NaN         1724.387271   
17207     Zimbabwe          ZWE  2022       NaN         2040.546587   
17208     Zimbabwe          ZWE  2023       NaN         2156.034093   
17209     Zimbabwe          ZWE  2024       NaN         2656.409377   

       edu_exp_pct_public  health_exp_pct_gdp  health_exp_pc_usd  \
17205                 NaN            2.954401          51.142506   
17206                 NaN            2.785717          63.511448   
17207                 NaN            3.628807          70.663139   
17208                 NaN                 NaN                NaN   
17209                 NaN                 NaN                NaN   

       infant_mortality  life_expectancy  fertility_rate  \
17205              44.9           61.530           3.754   
17206              43.6           60.135           3.765   
17207              42.2           62.360           3.767   
17208              40.6           62.775           3.724   
17209               NaN              NaN             NaN   

       age_dependency_ratio  population  
17205             84.384381  15526888.0  
17206             83.384953  15797210.0  
17207             82.547605  16069056.0  
17208             81.625352  16340822.0  
17209             80.100412  16634373.0

In [28]:
# save the dataset
df_world.to_csv('data/world_bank_data_cleaned.csv')

Explore other dataset.

In [29]:
# take a look into world happiness report
df_2015 = pd.read_csv('data/World Happiness Report/2015.csv')
df_2015.head(5)

Country          Region  Happiness Rank  Happiness Score  \
0  Switzerland  Western Europe               1            7.587   
1      Iceland  Western Europe               2            7.561   
2      Denmark  Western Europe               3            7.527   
3       Norway  Western Europe               4            7.522   
4       Canada   North America               5            7.427   

   Standard Error  Economy (GDP per Capita)   Family  \
0         0.03411                   1.39651  1.34951   
1         0.04884                   1.30232  1.40223   
2         0.03328                   1.32548  1.36058   
3         0.03880                   1.45900  1.33095   
4         0.03553                   1.32629  1.32261   

   Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
0                   0.94143  0.66557                        0.41978   
1                   0.94784  0.62877                        0.14145   
2                   0.87464  0.64938                        0.48357   
3                   0.88521  0.66973                        0.36503   
4                   0.90563  0.63297                        0.32957   

   Generosity  Dystopia Residual  
0     0.29678            2.51738  
1     0.43630            2.70201  
2     0.34139            2.49204  
3     0.34699            2.46531  
4     0.45811            2.45176

In [30]:
# take a look into world happiness report
df_2024 = pd.read_csv('data/World Happiness Report 2024/World-happiness-report-2024.csv')
df_2024.head(5)

Country name            Regional indicator  Ladder score  upperwhisker  \
0      Finland                Western Europe         7.741         7.815   
1      Denmark                Western Europe         7.583         7.665   
2      Iceland                Western Europe         7.525         7.618   
3       Sweden                Western Europe         7.344         7.422   
4       Israel  Middle East and North Africa         7.341         7.405   

   lowerwhisker  Log GDP per capita  Social support  Healthy life expectancy  \
0         7.667               1.844           1.572                    0.695   
1         7.500               1.908           1.520                    0.699   
2         7.433               1.881           1.617                    0.718   
3         7.267               1.878           1.501                    0.724   
4         7.277               1.803           1.513                    0.740   

   Freedom to make life choices  Generosity  Perceptions of corruption  \
0                         0.859       0.142                      0.546   
1                         0.823       0.204                      0.548   
2                         0.819       0.258                      0.182   
3                         0.838       0.221                      0.524   
4                         0.641       0.153                      0.193   

   Dystopia + residual  
0                2.082  
1                1.881  
2                2.050  
3                1.658  
4                2.298